### PR TITLE
drivers: watchdog: Add gpio based watchdog support

### PR DESCRIPTION
**Pull request by qianfan-Zhao**
Add gpio based watchdog(SP706) basic support.

Signed-off-by: qianfan Zhao <qianfanguijin@163.com>

### CODE REVIEW

File: drivers/watchdog/wdt_gpio.c [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+#include <watchdog.h>
+#include <gpio.h>
+
+#define LOG_LEVEL CONFIG_WDT_LOG_LEVEL
+#include <logging/log.h>
+LOG_MODULE_REGISTER(wdt_gpio);
+
+#define WDT_GPIO_PORT			WDT_GPIO_WDT_GPIO0_GPIO_CONTROLLER
+#define WDT_GPIO_PIN			WDT_GPIO_WDT_GPIO0_GPIO_PIN
+#define WDT_HW_MARGIN_MS		WDT_GPIO_WDT_GPIO0_HW_MARGIN_MS
```
Comment (qianfan-Zhao) [inline drivers/watchdog/wdt_gpio.c]: The marco name such as WDT_GPIO_WDT_GPIO0_GPIO_CONTROLLER are generated by using the node in dts, I don't think this is a good way, but don't know how to deal with this. Hope can get some feedbacks.

```
+	wdt-gpio0 {
+		compatible = "wdt-gpio";
+		label = "wdt-gpio0";
+		gpios = <&gpiob 21 0>;
+		hw-margin-ms = <1600>;
+		status = "ok";
+	};
```

File: drivers/watchdog/wdt_gpio.c [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+	k_mutex_init(&data->lock);
+	k_delayed_work_init(&data->work, wdt_gpio_delayed_work_timer);
+	k_delayed_work_submit(&data->work, WDT_HW_MARGIN_MS / 2);
+#endif
+
+	return 0;
+}
+
+static struct wdt_gpio_data private_data;
+DEVICE_AND_API_INIT(wdt_gpio, CONFIG_WDT_0_NAME, init_wdt,
```
Comment (qianfan-Zhao) [inline drivers/watchdog/wdt_gpio.c]: Wait the PR #11332 discuessed down.

CONFIG_WDT_0_NAME are defined in Kconfig, based on @nashif idea, this should be removed from Kconfig, generated in dts. So I will changed this.

File: dts/bindings/watchdog/wdt-gpio.yaml [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+      category: required
+      description: Human readable string describing the device (used by Zephyr for API name)
+      generation: define
+
+    gpios:
+      type: compound
+      category: required
+      generation: define
+
+    hw-margin-ms:
```
Comment (qianfan-Zhao) [inline dts/bindings/watchdog/wdt-gpio.yaml]: Is the name stype OK?

**Comment by qianfan-Zhao** [comment]
A sample dts patch to tested this PR:
```
diff --git a/boards/arm/nucleo_f070rb/nucleo_f070rb.dts b/boards/arm/nucleo_f070rb/nucleo_f070rb.dts
index 920304295a..d2ef957aab 100644
--- a/boards/arm/nucleo_f070rb/nucleo_f070rb.dts
+++ b/boards/arm/nucleo_f070rb/nucleo_f070rb.dts
@@ -17,14 +17,6 @@
 		zephyr,flash = &flash0;
 	};
 
-	leds {
-		compatible = "gpio-leds";
-		green_led_2: led_2 {
-			gpios = <&gpioa 5 GPIO_INT_ACTIVE_HIGH>;
-			label = "User LD2";
-		};
-	};
-
 	gpio_keys {
 		compatible = "gpio-keys";
 		user_button: button {
@@ -33,9 +25,19 @@
 		};
 	};
 
+	gpio_watchdogs {
+		compatible = "wdt-gpio";
+		watchdog0: wdt-gpio {
+			label = "wdt_gpio";
+			gpios = <&gpioa 5 GPIO_INT_ACTIVE_HIGH>;
+			hw-margin-ms = <1600>;
+			status = "ok";
+		};
+	};
+
 	aliases {
-		led0 = &green_led_2;
 		sw0 = &user_button;
+		wdt_gpio0 = &watchdog0;
 	};
 };
 ```

**Comment by qianfan-Zhao** [comment]
@ydamigos Could you please review the dts?

**Comment by nashif** [comment]
stale, closing. Reopen when there are updates...

**Pull request closed without merging**